crates/ast/src/lib.rs

```diff
@@ -0,0 +1,83 @@
+mod error;
+mod parser;
+
+pub use parser::parse;
+pub use error::Error;
+
+lalrpop_util::lalrpop_mod!(grammar);
+
+use alloy_dyn_abi::DynSolType;
+use alloy_primitives::U256;
+use revm_interpreter::opcode::OpCode;
+
+pub struct Root<'src>(pub Box<[HuffDefinition<'src>]>);
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum Instruction<'src> {
+    Op(OpCode),
+    Push(OpCode, U256),
+    PushAuto(U256),
+    LabelDefinition(&'src str),
+    LabelReference(&'src str),
+    MacroArgReference(&'src str),
+    ConstantReference(&'src str),
+    Invoke(Invoke<'src>),
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum Invoke<'src> {
+    Macro { name: &'src str, args: Box<[U256]> },
```

Comment:
A macro in huff can be invoked not only with constants but also opcodes and outer macro args